### PR TITLE
[llvm][DebugInfo] DWARFv5: static data members declarations are DW_TAG_variable

### PR DESCRIPTION
**Pull request by Michael137**
This patch adds the LLVM-side infrastructure to implement DWARFv5 issue 161118.1: "DW_TAG for C++ static data members".

The clang-side of this patch will simply construct the DIDerivedType with a different DW_TAG.

### CODE REVIEW

File: llvm/lib/IR/DebugInfo.cpp [hunk 1]
```diff
@@ -1315,17 +1315,15 @@ LLVMDIBuilderCreateUnspecifiedType(LLVMDIBuilderRef Builder, const char *Name,
...
 }
 
-LLVMMetadataRef
-LLVMDIBuilderCreateStaticMemberType(
+LLVMMetadataRef LLVMDIBuilderCreateStaticMemberType(
     LLVMDIBuilderRef Builder, LLVMMetadataRef Scope, const char *Name,
     size_t NameLen, LLVMMetadataRef File, unsigned LineNumber,
     LLVMMetadataRef Type, LLVMDIFlags Flags, LLVMValueRef ConstantVal,
-    uint32_t AlignInBits) {
+    unsigned Tag, uint32_t AlignInBits) {
```
Comment (adrian-prantl) [inline llvm/lib/IR/DebugInfo.cpp:1322]: Somebody might complain if we change the C API, but I don't think we guarantee compatibility.

**Review by adrian-prantl — APPROVED** [review]

**Review by adrian-prantl — CHANGES_REQUESTED** [review]
I think this is missing a test in clang/test/CodeGenCXX that verifies Clang generates the expected LLVM IR.

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 2]
```diff
@@ -1681,7 +1681,8 @@ CGDebugInfo::CreateRecordStaticField(const VarDecl *Var, llvm::DIType *RecordTy,
   llvm::DINode::DIFlags Flags = getAccessFlag(Var->getAccess(), RD);
   auto Align = getDeclAlignIfRequired(Var, CGM.getContext());
   llvm::DIDerivedType *GV = DBuilder.createStaticMemberType(
-      RecordTy, VName, VUnit, LineNumber, VTy, Flags, /* Val */ nullptr, Align);
+      RecordTy, VName, VUnit, LineNumber, VTy, Flags, /* Val */ nullptr,
+      llvm::dwarf::DW_TAG_member, Align);
```
Comment (adrian-prantl) [inline clang/lib/CodeGen/CGDebugInfo.cpp:1685]: Assuming this _changes_ the IR.

**Comment by Michael137** [comment]
> I think this is missing a test in clang/test/CodeGenCXX that verifies Clang generates the expected LLVM IR.

I added those tests here: https://github.com/llvm/llvm-project/pull/72235

The IR doesn't change with this patch. It merely adds the necessary parameter to the DIBuilder API to be used from clang in the subsequent patch

**Review by adrian-prantl — APPROVED** [review]
LGTM, then

**Comment by Michael137** [comment]
Looks like the OCaml bindings are using the C-API that was changed in this PR. Will make a follow-up fix for those

**Review by pogo59 — COMMENTED** [review]
@adrian-prantl We do have a pretty strong guarantee of compatibility in the llvm-c API. A new parameter pretty much means you need to define a new function.

**Comment by aeubanks** [comment]
we're seeing debug info verifier issues in Rust with this change, e.g.
```
 invalid tag
 !50 = !DIDerivedType(tag: DW_TAG_imported_declaration, name: "DISCR_EXACT", scope: !37, file: !2, baseType: !51, flags: DIFlagStaticMember, extraData: i64 0)
 invalid tag
 !61 = !DIDerivedType(tag: DW_TAG_imported_declaration, name: "DISCR_EXACT", scope: !54, file: !2, baseType: !51, flags: DIFlagStaticMember, extraData: i64 1)
 invalid tag
 !86 = !DIDerivedType(tag: DW_TAG_shared_type, name: "DISCR_EXACT", scope: !79, file: !2, baseType: !87, flags: DIFlagStaticMember, extraData: i64 0)
 invalid tag
 !97 = !DIDerivedType(tag: DW_TAG_shared_type, name: "DISCR_EXACT", scope: !90, file: !2, baseType: !87, flags: DIFlagStaticMember, extraData: i64 1)
```

is this expected?

(https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8764364541415601649/+/u/package_rust/stdout?format=raw)

**Comment by Michael137** [comment]
> we're seeing debug info verifier issues in Rust with this change, e.g.
> 
> ```
> 
>  invalid tag
> 
>  !50 = !DIDerivedType(tag: DW_TAG_imported_declaration, name: "DISCR_EXACT", scope: !37, file: !2, baseType: !51, flags: DIFlagStaticMember, extraData: i64 0)
> 
>  invalid tag
> 
>  !61 = !DIDerivedType(tag: DW_TAG_imported_declaration, name: "DISCR_EXACT", scope: !54, file: !2, baseType: !51, flags: DIFlagStaticMember, extraData: i64 1)
> 
>  invalid tag
> 
>  !86 = !DIDerivedType(tag: DW_TAG_shared_type, name: "DISCR_EXACT", scope: !79, file: !2, baseType: !87, flags: DIFlagStaticMember, extraData: i64 0)
> 
>  invalid tag
> 
>  !97 = !DIDerivedType(tag: DW_TAG_shared_type, name: "DISCR_EXACT", scope: !90, file: !2, baseType: !87, flags: DIFlagStaticMember, extraData: i64 1)
> 
> ```
> 
> 
> 
> is this expected?
> 
> 
> 
> (https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8764364541415601649/+/u/package_rust/stdout?format=raw)
> 
> 

Hmmm yea that's problematic. Feel free to revert since I'm currently not at my PC. It looks like the Tag parameter wouldve been better off strongly typed since I suspect there are some callers that I missed that now pass alignment to where now the tag is.

**Comment by aeubanks** [comment]
ah I see, that definitely sounds plausible

I've also recently done some NFC changes to better catch issues like this in https://github.com/llvm/llvm-project/pull/66295

**Comment by aeubanks** [comment]
actually I think the caller is in Rust, so perhaps we should just update Rust instead. but making some of these params more strongly typed would be good

**Comment by Michael137** [comment]
Is it possible that this is to be fixed at the callsites in the rust frontend? Dont think i missed any callers in LLVM (then again I'm not at a PC to confirm this atm)

**Pull request merged**